File: src/strategies/orderbook-imbalance.ts
/**
 * Order Book Imbalance Strategy (NEW — not in v1)
 *
 * Thesis: When the order book has a heavily imbalanced bid/ask wall within
 * 1% of the current price, it creates a magnetic pull — price tends to move
 * toward large walls (support/resistance) before blowing through them.
 *
 * Two plays:
 *  A. Bid wall support: large bids stacked below price → buy the support
 *  B. Ask wall absorption: price pressing into a large ask wall →
 *     wait for the wall to be absorbed (eaten), then ride the breakout
 *
 * This requires L2 order book data (Coinbase Advanced WebSocket).
 */

import type { TradeSignal, ScannerConfig } from '../types.js';
import { randomUUID } from 'crypto';

interface OrderBookLevel {
  price: number;
  size: number; // in base currency
}

interface OrderBookSnapshot {
  bids: OrderBookLevel[]; // sorted descending
  asks: OrderBookLevel[]; // sorted ascending
  lastUpdated: number;
}

const orderBooks = new Map<string, OrderBookSnapshot>();

export function updateOrderBook(symbol: string, bids: OrderBookLevel[], asks: OrderBookLevel[]): void {
  orderBooks.set(symbol, {
    bids: [...bids].sort((a, b) => b.price - a.price),
    asks: [...asks].sort((a, b) => a.price - b.price),
    lastUpdated: Date.now(),
  });
}

function sumBookDepth(levels: OrderBookLevel[], from: number, pricePct: number): number {
  const limit = from * (1 + pricePct);
  return levels
    .filter(l => Math.abs(l.price - from) / from <= pricePct)
    .reduce((s, l) => s + l.size * l.price, 0);
}

export function scanOrderBookImbalance(
  symbol: string,
  productId: string,
  currentPrice: number,
  config: ScannerConfig,
): TradeSignal | null {
  const book = orderBooks.get(symbol);
  if (!book || Date.now() - book.lastUpdated > 30_000) return null; // stale

  // Measure bid/ask depth within 1% of price
  const bidDepthUsd = sumBookDepth(book.bids, currentPrice, 0.01);
  const askDepthUsd = sumBookDepth(book.asks, currentPrice, 0.01);
  const totalDepth = bidDepthUsd + askDepthUsd;

  if (totalDepth < 500_000) return null; // not enough liquidity to signal

  const imbalanceRatio = bidDepthUsd / (askDepthUsd + 1); // >1 = bid heavy

  // ── Bid wall support play ─────────────────────────────────────────────
  if (imbalanceRatio > 3.0 && bidDepthUsd > 2_000_000) {
    const wallScore = Math.min(30, imbalanceRatio * 5);
    const sizeScore = Math.min(20, bidDepthUsd / 200_000);
    const score = Math.min(82, 42 + wallScore + sizeScore);

    if (score >= config.minQualScoreScalp) {
      return {
        id: randomUUID(),
        symbol,
        productId,
        strategy: 'orderbook_imbalance',
        side: 'long',
        tier: 'scalp',
        score,
        confidence: 'low',
        sources: ['orderbook'],
        reasoning: `${symbol} bid/ask ratio ${imbalanceRatio.toFixed(1)}× within 1%, $${(bidDepthUsd / 1e6).toFixed(1)}M bid wall support`,
        entryPrice: currentPrice,
        stopPrice: currentPrice * 0.99,
        suggestedSizeUsd: 40,
        expiresAt: Date.now() + 300_000, // 5 min
        createdAt: Date.now(),
      };
    }
  }

  // ── Ask wall short (distribution wall) ───────────────────────────────
  if (imbalanceRatio < 0.33 && askDepthUsd > 2_000_000) {
    const wallScore = Math.min(25, (1 / (imbalanceRatio + 0.01)) * 3);
    const sizeScore = Math.min(20, askDepthUsd / 200_000);
    const score = Math.min(80, 40 + wallScore + sizeScore);

    if (score >= config.minQualScoreScalp) {
      return {
        id: randomUUID(),
        symbol,
        productId,
        strategy: 'orderbook_imbalance',
        side: 'short',
        tier: 'scalp',
        score,
        confidence: 'low',
        sources: ['orderbook'],
        reasoning: `${symbol} ask/bid ratio ${(1 / imbalanceRatio).toFixed(1)}× within 1%, $${(askDepthUsd / 1e6).toFixed(1)}M ask wall resistance`,
        entryPrice: currentPrice,
        stopPrice: currentPrice * 1.01,
        suggestedSizeUsd: 35,
        expiresAt: Date.now() + 300_000,
        createdAt: Date.now(),
      };
    }
  }

  return null;
}

File: src/strategies/orderbook_imbalance.py
"""Order Book Imbalance Strategy."""

import time
import uuid
from dataclasses import dataclass, field
from typing import Optional

from src.types import TradeSignal, ScannerConfig


@dataclass
class OrderBookLevel:
    price: float
    size: float


@dataclass
class OrderBookSnapshot:
    bids: list[OrderBookLevel] = field(default_factory=list)
    asks: list[OrderBookLevel] = field(default_factory=list)
    last_updated: float = 0


_order_books: dict[str, OrderBookSnapshot] = {}
_MAX_ORDER_BOOKS = 500


def update_order_book(symbol: str, bids: list[OrderBookLevel], asks: list[OrderBookLevel]) -> None:
    if symbol not in _order_books and len(_order_books) >= _MAX_ORDER_BOOKS:
        oldest_key = min(_order_books, key=lambda k: _order_books[k].last_updated)
        del _order_books[oldest_key]
    _order_books[symbol] = OrderBookSnapshot(
        bids=sorted(bids, key=lambda l: l.price, reverse=True),
        asks=sorted(asks, key=lambda l: l.price),
        last_updated=time.time() * 1000,
    )


def _sum_book_depth(levels: list[OrderBookLevel], from_price: float, price_pct: float) -> float:
    if from_price == 0:
        return 0.0
    return sum(
        l.size * l.price for l in levels
        if abs(l.price - from_price) / from_price <= price_pct
    )


def get_bid_ask_spread(symbol: str) -> Optional[float]:
    """Return the bid-ask spread as a fraction of mid price, or None if no book data.

    E.g., 0.005 means 0.5% spread.
    """
    book = _order_books.get(symbol)
    if not book or not book.bids or not book.asks:
        return None
    now = time.time() * 1000
    if now - book.last_updated > 30_000:
        return None  # stale book data
    best_bid = book.bids[0].price
    best_ask = book.asks[0].price
    if best_bid <= 0 or best_ask <= 0:
        return None
    mid = (best_bid + best_ask) / 2
    return (best_ask - best_bid) / mid


def scan_orderbook_imbalance(
    symbol: str, product_id: str, current_price: float,
    config: ScannerConfig,
) -> Optional[TradeSignal]:
    book = _order_books.get(symbol)
    now = time.time() * 1000
    if not book or now - book.last_updated > 30_000:
        return None

    bid_depth = _sum_book_depth(book.bids, current_price, 0.01)
    ask_depth = _sum_book_depth(book.asks, current_price, 0.01)
    total_depth = bid_depth + ask_depth
    if total_depth < 500_000:
        return None

    imbalance_ratio = bid_depth / (ask_depth + 1)

    # Bid wall support
    if imbalance_ratio > 3.0 and bid_depth > 2_000_000:
        wall_score = min(30, imbalance_ratio * 5)
        size_score = min(20, bid_depth / 200_000)
        score = min(82, 42 + wall_score + size_score)
        if score >= config.min_qual_score_scalp:
            return TradeSignal(
                id=str(uuid.uuid4()), symbol=symbol, product_id=product_id,
                strategy="orderbook_imbalance", side="long", tier="scalp", score=score,
                confidence="low", sources=["orderbook"],
                reasoning=f"{symbol} bid/ask ratio {imbalance_ratio:.1f}x within 1%, ${bid_depth/1e6:.1f}M bid wall",
                entry_price=current_price, stop_price=current_price * 0.99,
                suggested_size_usd=40,
                expires_at=now + 300_000, created_at=now,
            )

    # Ask wall short
    if imbalance_ratio < 0.33 and ask_depth > 2_000_000:
        wall_score = min(25, (1 / (imbalance_ratio + 0.01)) * 3)
        size_score = min(20, ask_depth / 200_000)
        score = min(80, 40 + wall_score + size_score)
        if score >= config.min_qual_score_scalp:
            return TradeSignal(
                id=str(uuid.uuid4()), symbol=symbol, product_id=product_id,
                strategy="orderbook_imbalance", side="short", tier="scalp", score=score,
                confidence="low", sources=["orderbook"],
                reasoning=f"{symbol} ask/bid ratio {1/max(imbalance_ratio, 0.001):.1f}x within 1%, ${ask_depth/1e6:.1f}M ask wall",
                entry_price=current_price, stop_price=current_price * 1.01,
                suggested_size_usd=35,
                expires_at=now + 300_000, created_at=now,
            )

    return None
